[JENKINS-55853] Fix post build comments for matrix builds

proski opened the pull request:
```
*  Fix post build comments for matrix builds
   
   Process post build comments in StashBuilds#onCompleted(), which is run in
   the top-level build, unlike StashPostBuildComment#perform() that is run
   in the child builds.
   
   Make StashPostBuildComment implement Describable<Publisher>. Find the
   StashPostBuildComment object in the publisher list of the project.
   If the post build comment cannot be expanded, post it unexpanded with a
   warning.
   
   Keep StashPostBuildCommentAction for backward compatibility.
```

proski commented:
Resolves https://issues.jenkins-ci.org/browse/JENKINS-55853
Tested on Jenkins 2.164 on free-style and matrix builds with and without post build comments.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuilds.java:
```diff
@@ -64,12 +65,21 @@ public void onCompleted(AbstractBuild build, TaskListener listener) {
+                String expandedComment;
+                try {
+                    expandedComment = Util.fixEmptyAndTrim(build.getEnvironment(listener).expand(buildComment));
+                } catch (Exception e) {
```
Can we get rid of this? `expand()` doesn't declare any checked exceptions, and catching `Exception` is generally discouraged

proski replied:
`AbstractBuild#getEnvironment()` throws IOException and InterruptedException, so they must be caught. I agree that catching `Exception` is bad, I copied that from `StashPostBuildComment#perform()` without giving it a thought. Fixing it now.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashPostBuildCommentAction.java:
```diff
@@ -1,21 +0,0 @@
-package stashpullrequestbuilder.stashpullrequestbuilder;
```
I'm not sure, but won't removing this class cause warnings about unrecognized data stored in old builds?

proski replied:
Actually yes. It started the next day after I installed the plugin with this fix. No idea why it's taking Jenkins so long. That said, I don't think we should keep unused code to prevent those warnings. If there is a nice way to tell Jenkins to suppress those warnings (like "that class is obsolete, don't worry"), I would be happy to incorporate it.

jakub-bochenski replied:
I think we need to do a data migration for this, but I'm not sure about specifics. Best I can offer for now is https://wiki.jenkins.io/display/JENKINS/Hint+on+retaining+backward+compatibility#Hintonretainingbackwardcompatibility-Scenario:renameaclass but this is for renaming a class, not removing it

proski replied:
Good, thank you! I think it can be added to `StashPostBuildComment`. That would take some time to test. I'll add a commit for that once I'm sure it's working.

proski replied:
OK, I wrote the code, but it doesn't work, and I know why. The `build.xml` files contain information file this:
```
<build>
  <action>
    <stashpullrequestbuilder.stashpullrequestbuilder.StashPostBuildCommentAction plugin="stash-pullrequest-builder@1.9-SNAPSHOT">
        <buildSuccessfulComment>Freestyle is good for 054c04bd78511efd70b6ad82c8c09f973f61bb62</buildSuccessfulComment>
        <buildFailedComment>Freestyle is bad for 054c04bd78511efd70b6ad82c8c09f973f61bb62</buildFailedComment>
    </stashpullrequestbuilder.stashpullrequestbuilder.StashPostBuildCommentAction>
  </action>
</build>
```
Once this PR is applied, there is no other Action that can read that data. StashPostBuildComment is not an Action.

Is that data useful? I don't think so. It's what the plugin would report in case of success and failure. It doesn't tell the user what _actually_ happened. That's a bad design, as it calculates data that may not be needed (I'm actually thinking of adding messages for unstable and aborted builds - unstable builds are actually the ones I want to explain to my users in detail).

I guess we should just accept that the warning would appear. It should be a one time thing. Jenkins will discard the data that should not have been recorded in the first place.

jakub-bochenski replied:
> I guess we should just accept that the warning would appear. It should be a one time thing. Jenkins will discard the data that should not have been recorded in the first place.

I agree, cleaning up the build.xml should be an extra benefit

proski commented:
Just a rebase on top of the current master branch for everybody's convenience.

proski commented:
Please consider this PR on hold until further notice. It's good to have an `InvisibleAction` to hold messages from the trigger, such as warnings about misconfiguration.

proski commented:
OK, this PR is no longer on hold. I restored `StashPostBuildCommentAction.java` and the fields that were there. The fields are now `transient` so they can be loaded, but won't be saved. That fixes the warnings about unreadable data.

If we need to pass any messages from the the trigger to the build console output or to Stash comments, adding them to an InvisibleAction is probably the right thing to do. I'm not going to try it in this PR.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashPostBuildCommentAction.java:
```diff
@@ -3,19 +3,10 @@
 import hudson.model.InvisibleAction;
 
 public class StashPostBuildCommentAction extends InvisibleAction {
```
Maybe we should rename the class since it's not longer about build comments? (Using that rename class migration linked above)

proski replied:
The only purpose of that class now is to be a compatibility data consumer/destroyer of its former self, i.e. `StashPostBuildCommentAction`. I cannot think of a better name for it.

jakub-bochenski replied:
OK, nevermind then

proski commented:
I started fixing warnings using Eclipse, and discovered that some of the warnings are added by this PR. So I'm adding a fixup commit here to fix those warnings.

proski commented:
Please consider this PR on hold until I retest it. I want to make sure that `onCompleted` would still run with the new signature.

proski commented:
No longer on hold. The testing was successful. Rebased and force-pushed to trigger the rebuild.

jakub-bochenski commented:
@jimklimov it would be nice to merge this and other PRs that are already reviewed. If you can't review them soon let's merge them. If you find something to be problematic we can reopen the case

proski commented:
I'll rebase my PRs later today (California time).

proski committed:
Fix post build comments for matrix builds

Process post build comments in StashBuilds#onCompleted(), which is run in
the top-level build, unlike StashPostBuildComment#perform() that is run
in the child builds.

Make StashPostBuildComment implement Describable<Publisher>. Find the
StashPostBuildComment object in the publisher list of the project.
If the post build comment cannot be expanded, post it unexpanded with a
warning.

Keep StashPostBuildCommentAction for backward compatibility.
